test/base-fork/SparkVault.t.sol

```diff
@@ -0,0 +1,193 @@
+        vm.stopPrank();
+
+        key = RateLimitHelpers.makeAssetKey(
+            LIMIT_SPARK_VAULT_TAKE,
+            address(sparkVault)
+        );
+
+        vm.startPrank(Base.SPARK_EXECUTOR);
+        rateLimits.setRateLimitData(key, 1_000_000e18, uint256(1_000_000e18) / 1 days);
+        vm.stopPrank();
+    }
+
+    function _assertTestState(TestState memory state, uint256 tolerance) internal view {
+        assertApproxEqAbs(rateLimits.getCurrentRateLimit(key),  state.rateLimit,        tolerance, "rateLimit");
+        assertApproxEqAbs(asset.balanceOf(address(this)),       state.assetThis,        tolerance, "assetThis");
+        assertApproxEqAbs(asset.balanceOf(address(almProxy)),   state.assetAlm,         tolerance, "assetAlm");
+        assertApproxEqAbs(asset.balanceOf(address(sparkVault)), state.assetVault,       tolerance, "assetVault");
+        assertApproxEqAbs(sparkVault.balanceOf(address(this)),  state.vaultThis,        tolerance, "vaultThis");
+        assertApproxEqAbs(sparkVault.totalAssets(),             state.vaultTotalAssets, tolerance, "vaultTotalAssets");
+        assertApproxEqAbs(sparkVault.totalSupply(),             state.vaultTotalSupply, tolerance, "vaultTotalSupply");
+    }
+
+    function _assertTestState(TestState memory state) internal view {
+        _assertTestState(state, 0);
+    }
+}
+
+contract ForeignControllerTakeFromSparkVaultFailureTests is ForeignControllerTakeFromSparkVaultTestBase {
+
+    function test_takeFromSparkVault_notRelayer() external {
+        vm.expectRevert(abi.encodeWithSignature(
+            "AccessControlUnauthorizedAccount(address,bytes32)",
+            address(this),
+            RELAYER
+        ));
+        foreignController.takeFromSparkVault(address(sparkVault), 1e18);
+    }
+
+    function test_takeFromSparkVault_zeroMaxAmount() external {
+        vm.startPrank(Base.SPARK_EXECUTOR);
+        rateLimits.setRateLimitData(key, 0, 0);
+        vm.stopPrank();
+
+        vm.prank(relayer);
+        vm.expectRevert("RateLimits/zero-maxAmount");
+        foreignController.takeFromSparkVault(address(sparkVault), 1e18);
+    }
+
+    function test_takeFromSparkVault_rateLimitBoundary() external {
+        deal(address(asset), address(this), 10_000_000e18);
+        asset.approve(address(sparkVault), 10_000_000e18);
+        sparkVault.mint(10_000_000e18, address(this));
+
+        vm.startPrank(Base.SPARK_EXECUTOR);
+        rateLimits.setRateLimitData(key, 10_000_000e18, uint256(10_000_000e18) / 1 days);
+        vm.stopPrank();
+
+        vm.prank(relayer);
+        vm.expectRevert("RateLimits/rate-limit-exceeded");
+        foreignController.takeFromSparkVault(address(sparkVault), 10_000_000e18 + 1);
+
+        vm.prank(relayer);
+        foreignController.takeFromSparkVault(address(sparkVault), 10_000_000e18);
+    }
+
+}
+
+contract ForeignControllerTakeFromSparkVaultTests is ForeignControllerTakeFromSparkVaultTestBase {
+
```

Comment:
Pattern used is to make a user `address user = makeAddr("user");` and use this for balance assertions etc